clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4570,6 +5660,33 @@ void Sema::MarkDevices() {
   }
 }
 
+void Sema::ProcessFreeFunctions() {
+  for (Decl *D : syclDeviceDecls()) {
+    auto *SYCLKernel = cast<FunctionDecl>(D);
+    if (IsFreeFunction(*this, SYCLKernel)) {
+      SyclKernelDecompMarker DecompMarker(*this);
+      SyclKernelFieldChecker FieldChecker(*this);
+      SyclKernelUnionChecker UnionChecker(*this);
+
+      KernelObjVisitor Visitor{*this};
+
+      DiagnosingSYCLKernel = true;
+
+      // Check parameters of free function.
+      Visitor.VisitFunctionParameters(SYCLKernel, DecompMarker, FieldChecker,
+                                      UnionChecker);
+
+      DiagnosingSYCLKernel = false;
+
+      // Ignore the free function if any of the checkers fail validation.
+      if (!FieldChecker.isValid() || !UnionChecker.isValid())
+        return;
+
+      ConstructFreeFunctionKernel(*this, SYCLKernel);
```

Comment:
I'm not sure if end of TU is the right place to handle this. Did you try calling this earlier? Maybe at ActOnFinishFunctionBody and PerformPendingInstantations.

Reply:
Let me try moving it to ActOnFinishFunctionBody. That will avoid the scan of all funtions searching for a free function.

Reply:
If I try to process a free function at ActOnFinishFunctionBody then templated free functions don't work because their instantiation does not appear to be done. If I were to try to process free functions at PerformPendingInstantations then non-templated free functions would not get caught.
Is there a place when a function has been seen, templated or non-templated, where I can then check for free function attributes and process it?

Reply:
Hmm.. I am not sure. I think you may need to call it both places. @AaronBallman can you weigh in?

Reply:
`ActOnFinishFunctionBody()` should get called twice (or more). Once when acting on the templated body and once again when that body is instantiated, right?